### ISSUE TITLE
Compression: etag wrong

### ISSUE THREAD

**Issue by Tarrowren** [minio/minio#12337]
## Current Behavior

I have compression enabled for PDF
```
mc admin config set minio-test compression extensions=".pdf" mime_types="application/pdf"
```
upload PDF file using the AWS SDK，
exception:  
```
software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Invalid base 16 character: '-'
```

but normal when compression disabled.

## Environment

minio version RELEASE.2021-04-22T15-44-28Z
aws sdk 2.16.66

**Comment by klauspost** [comment]
Ref https://github.com/minio/minio/pull/11577

TLDR; is that etag may or may not be the content hash.

@aead Do you remember the reason for not having the etag be the content hash (without -1) on compression only?

**Comment by harshavardhana** [comment]
> Ref #11577
> 
> TLDR; is that etag may or may not be the content hash.
> 
> @aead Do you remember the reason for not having the etag be the content hash (without -1) on compression only?

@klauspost @aead we will have to add -1 here this might also automatically fix the mint functional tests with compression + encryption

**Comment by klauspost** [comment]
@harshavardhana ?  `-1` is always added to compressed objects, since forever. That is why the AWS SDK above is failing. It expects it to be base 16(hex) and not contain `-`.

**Comment by aead** [comment]
We should not add the `-1` prefix for single part objects - since that is the multipart behavior.
When compressing an object - in case of single part - we should simply compute the ETag as `MD5(object)`.

**Comment by aead** [comment]
@klauspost I don't recall why we add `-1` but we compute the regular MD5 etag **and then** append a `-1` prefix. We could simply stop doing this unless there is a specific reason for this.

**Comment by harshavardhana** [comment]
> @klauspost I don't recall why we add `-1` but we compute the regular MD5 etag **and then** append a `-1` prefix. We could simply stop doing this unless there is a specific reason for this.

-1 is added to avoid clients interpreting the ETag as md5sum @aead @klauspost - such as s3cmd ignores `-1` 

> When compressing an object - in case of single part - we should simply compute the ETag as `MD5(object)`.

If this is expected currently `compression` enabled doesn't return the actual etag of the content.

**Comment by harshavardhana** [comment]
> @harshavardhana ? `-1` is always added to compressed objects, since forever. That is why the AWS SDK above is failing. It expects it to be base 16(hex) and not contain `-`.

@klauspost correct that is expected since we can't calculate etag properly for compressed objects due to the streaming approach. 

My assumption was after the pkg/etag introduction this would be handled properly

**Comment by aead** [comment]
> -1 is added to avoid clients interpreting the ETag as md5sum @aead @klauspost - such as s3cmd ignores -1

@harshavardhana My point is that an ETag of a compressed object should not differ from the ETag of an uncompressed object. There is no reason (AFAICS) to leak (through S3 API) that we treat the object differently.

> @klauspost correct that is expected since we can't calculate etag properly for compressed objects due to the streaming approach.

No, we should be able to do that. The ETag computation is happening before compression is applied.
See:
 - https://github.com/minio/minio/blob/master/cmd/object-handlers.go#L1583
 - https://github.com/minio/minio/blob/master/cmd/object-handlers.go#L1592
 -  https://github.com/minio/minio/blob/abd32065aaae4080396a1b4b04a110454368b028/pkg/hash/reader.go#L121

**Comment by aead** [comment]
It seems like there is some issue in the current ETag computation construction. I will investigate and try to unify ETags for compressed and uncompressed single-part objects...

**Comment by harshavardhana** [comment]
> It seems like there is some issue in the current ETag computation construction. I will investigate and try to unify ETags for compressed and uncompressed single-part objects... 

Okay @aead

**Comment by harshavardhana** [comment]
Closing as duplicate of #11447
